GitUI/GitUIExtensions.cs

```diff
@@ -117,11 +117,22 @@ public static Task ViewChangesAsync(this FileViewer diffViewer, IReadOnlyList<Gi
                 return diffViewer.ViewPatchAsync(() =>
                 {
                     string selectedPatch = diffViewer.GetSelectedPatch(firstRevision, secondRevision, file);
-                    return selectedPatch ?? defaultText;
+                    if (selectedPatch == null)
+                    {
+                        return Tuple.Create(defaultText, CannotOpenWithDifftool);
+                    }
+                    else
+                    {
+                        return Tuple.Create<string, Action>(selectedPatch,
+                            () => { diffViewer.Module.OpenWithDifftool(file.Name, null, firstRevision, secondRevision, "", file.IsTracked); });
+                    }
                 });
             }
         }
 
+        public const Action CannotOpenWithDifftool = null;
+        public const Action OpenWithDifftoolNotImplemented = null;
```

Comment:
RE: FileTree - the file tree displays the state of the selected commit and *not* the state of the working directory, unless the HEAD commit is selected.
So pressing F3 should open a diff tool showing a diff between the current working folder version and the version in the selected commit. If HEAD is selected then diff would be nothing, which is expected.

Thoughts?
/cc: @gerhardol @jbialobr

Reply:
> I wanted the new argument [CanBeNull] Action openWithDifftool of the function FileViewer.ViewPatch() to be mandatory in order to adapt all its callers. For some of them this action really doesn't apply. Usually a default value null could be used in this case. That's why I created the named null CannotOpenWithDiffTool. In other cases there is displayed a real diff but I didn't know how to and/or didn't find it important to open it in the external diff tool --> OpenWithDifftoolNotImplemented.

Thank you for the clarification.
I suppose it can be done with a proxy type, however I personally don't think this distinction is particularly important in this case - the end result we either open a diff tool, or we don't.
In each case where we pass `null` we can have a comment to explain _why_ - e.g. not applicable or not important.

Reply:
>Thoughts?

Sounds like a good idea.